### PR TITLE
fix: boosting syncing process

### PR DESCRIPTION
**Pull request by b00f**
## Description

This PR aims to enhance the syncing process by implementing the following changes:

- Weighting peers in random selection based on their success rate.
- Closing the session when a connection cannot be established.
- The default session timeout decreased to 10 seconds.

These modifications are expected to improve the overall speed of the syncing process.

### CODE REVIEW

File: sync/peerset/peer_set.go [hunk 1]
```diff
@@ -31,23 +32,11 @@ func NewPeerSet(sessionTimeout time.Duration) *PeerSet {
...
-	}
-
-	return Peer{}
-}
-
 func (ps *PeerSet) OpenSession(pid peer.ID) *Session {
 	ps.lk.Lock()
 	defer ps.lk.Unlock()
 
+	//TODO: check if pid exists
```
Comment (themantre) [inline sync/peerset/peer_set.go:39]: Remove this comment or update OpenSession calls to make sure it is not nil

File: sync/sync.go [hunk 2]
```diff
@@ -294,15 +297,22 @@ func (sync *synchronizer) prepareBundle(msg message.Message) *bundle.Bundle {
...
 
-func (sync *synchronizer) sendTo(msg message.Message, to peer.ID) {
+func (sync *synchronizer) sendTo(msg message.Message, to peer.ID, sessionID int) {
 	bdl := sync.prepareBundle(msg)
 	if bdl != nil {
 		data, _ := bdl.Encode()
 		err := sync.network.SendTo(data, to)
 		if err != nil {
 			sync.logger.Error("error on sending bundle", "bundle", bdl, "err", err, "to", to)
+			sync.peerSet.IncreaseSendFailedCounter(to)
```
Comment (themantre) [inline sync/sync.go:307]: Should we increase the failed attempts when the connection is OK but request rejectes?

**Review by themantre — APPROVED** [review]

File: sync/peerset/peer_set.go (hunk 1 above)
Comment (b00f) [reply sync/peerset/peer_set.go:39]: Better to remove the comment

File: sync/sync.go (hunk 2 above)
Comment (b00f) [reply sync/sync.go:310]: It's done

File: sync/handler_blocks_response_test.go [hunk 3]
```diff
@@ -122,27 +121,38 @@ func TestSyncing(t *testing.T) {
...
+	shouldPublishMessageWithThisType(t, networkBob, message.MessageTypeBlocksResponse) // 70-80
+	shouldPublishMessageWithThisType(t, networkBob, message.MessageTypeBlocksResponse) // 81-91
+	shouldPublishMessageWithThisType(t, networkBob, message.MessageTypeBlocksResponse) // 92-92
 	shouldPublishMessageWithThisType(t, networkBob, message.MessageTypeBlocksResponse) // NoMoreBlock
 
 	// Latest block requests
 	shouldPublishMessageWithThisType(t, networkAlice, message.MessageTypeBlocksRequest)
-	shouldPublishMessageWithThisType(t, networkBob, message.MessageTypeBlocksResponse) // 91-100
+	shouldPublishMessageWithThisType(t, networkBob, message.MessageTypeBlocksResponse) // 93-100
 	shouldPublishMessageWithThisType(t, networkBob, message.MessageTypeBlocksResponse) // Synced
```
Comment (themantre) [inline sync/handler_blocks_response_test.go:154]: check this bundle to make sure it is synced

**Review by themantre — CHANGES_REQUESTED** [review]

File: sync/bundle/message/blocks_response.go [hunk 4]
```diff
@@ -33,6 +34,9 @@ func (m *BlocksResponseMessage) SanityCheck() error {
 			return err
 		}
 	}
+	if m.From == 0 && len(m.Blocks) != 0 {
+		return errors.Errorf(errors.ErrInvalidHeight, "unexpected blocks for height zero")
```
[the remark this answers is not in the document]
Comment (b00f) [reply sync/bundle/message/blocks_response.go:38]: Done

File: sync/handler_blocks_response_test.go (hunk 3 above)
Comment (b00f) [reply sync/handler_blocks_response_test.go:154]: Done

File: sync/bundle/message/blocks_response_test.go [hunk 5]
```diff
@@ -14,20 +15,29 @@ func TestLatestBlocksResponseType(t *testing.T) {
...
 	t.Run("Invalid certificate", func(t *testing.T) {
 		b := block.GenerateTestBlock(nil, nil)
 		c := block.NewCertificate(-1, nil, nil, nil)
-		m := NewBlocksResponseMessage(ResponseCodeMoreBlocks, 1, 100, []*block.Block{b}, c)
+		m := NewBlocksResponseMessage(ResponseCodeMoreBlocks, sid, 100, []*block.Block{b}, c)
 
 		assert.Equal(t, errors.Code(m.SanityCheck()), errors.ErrInvalidRound)
 	})
 
+	t.Run("From height set to zero with a block", func(t *testing.T) {
```
Comment (themantre) [inline sync/bundle/message/blocks_response_test.go:27]:
```suggestion
	t.Run("From set to zero with a block", func(t *testing.T) {
```

**Review by themantre — APPROVED** [review]

**Review by themantre — APPROVED** [review]

**Pull request merged**